include/cliffordsynthesis/encoding/ObjectiveEncoder.hpp

```diff
@@ -0,0 +1,98 @@
+//
+// This file is part of the MQT QMAP library released under the MIT license.
+// See README.md or go to https://github.com/cda-tum/qmap for more information.
+//
+
+#pragma once
+
+#include "cliffordsynthesis/TargetMetric.hpp"
+#include "cliffordsynthesis/encoding/GateEncoder.hpp"
+#include "cliffordsynthesis/encoding/TableauEncoder.hpp"
+
+#include <cstddef>
+#include <functional>
+#include <memory>
+
+namespace cs::encoding {
+
+class ObjectiveEncoder {
+public:
+  ObjectiveEncoder(const std::size_t nQubits, const std::size_t timestepLimit,
+                   TableauEncoder::Variables*                    tvars,
+                   GateEncoder::Variables*                       gvars,
+                   const std::shared_ptr<logicbase::LogicBlock>& lb)
+      : N(nQubits), T(timestepLimit), tvars(tvars), gvars(gvars), lb(lb) {}
+
+  template <class Op>
+  void limitGateCount(std::size_t maxGateCount, Op op,
+                      bool includeSingleQubitGates = true) const {
+    DEBUG() << "Limiting gate count to at most " << maxGateCount
+            << (includeSingleQubitGates ? "" : " two-qubit") << " gate(s)";
+
+    const auto cost = collectGateCount(includeSingleQubitGates);
+    lb->assertFormula(
+        op(cost, logicbase::LogicTerm(static_cast<int>(maxGateCount))));
+  }
+
+  void optimizeMetric(TargetMetric targetMetric) const;
+
+  void optimizeGateCount(bool includeSingleQubitGates = true) const;
+
+  void optimizeDepth(bool includeSingleQubitGates = true) const;
+
+protected:
+  // number of qubits N
+  std::size_t N{};
+  // timestep limit T
+  std::size_t T{};
+
+  // the tableau variables
+  TableauEncoder::Variables* tvars;
```

Comment:
These are not actually needed, are they?
I cannot think of a objective that needs to access the Tableau Variables.

Reply:
👍🏼 Fixed in https://github.com/cda-tum/qmap/pull/187/commits/26095623fab34fdd0d1296c9977766ceab86bc84.